server/api.go

```diff
@@ -0,0 +1,82 @@
+package main
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"github.com/mattermost/mattermost-plugin-msteams-sync/server/msteams"
+)
+
+type API struct {
+	p      *Plugin
+	router *mux.Router
+}
+
+type Activities struct {
+	Value []msteams.Activity
+}
+
+func NewAPI(p *Plugin) *API {
+	router := mux.NewRouter()
+	api := &API{p: p, router: router}
+
+	router.HandleFunc("/avatar/{userId:.*}", api.getAvatar).Methods("GET")
+	router.HandleFunc("/", api.processMessage).Methods("POST")
+
+	return api
+}
+
+func (a *API) getAvatar(w http.ResponseWriter, r *http.Request) {
+	params := mux.Vars(r)
+	userID := params["userId"]
+	photo, appErr := a.p.API.KVGet("avatar_" + userID)
+	if appErr != nil || len(photo) == 0 {
+		var err error
+		photo, err = a.p.msteamsAppClient.GetUserAvatar(userID)
+		if err != nil {
+			a.p.API.LogError("Unable to read avatar", "error", err)
+			http.Error(w, "avatar not found", http.StatusNotFound)
+			return
+		}
+
+		appErr := a.p.API.KVSetWithExpiry("avatar_"+userID, photo, 300)
+		if appErr != nil {
+			a.p.API.LogError("Unable to cache the new avatar", "error", appErr)
+			return
+		}
+	}
+	w.Write(photo)
+}
+
+func (a *API) processMessage(w http.ResponseWriter, req *http.Request) {
+	validationToken := req.URL.Query().Get("validationToken")
+	if validationToken != "" {
+		w.Write([]byte(validationToken))
+		return
+	}
+
+	activities := Activities{}
+	err := json.NewDecoder(req.Body).Decode(&activities)
+	if err != nil {
+		http.Error(w, "unable to get the activities from the message", http.StatusBadRequest)
```

Comment:
We should log this error

Reply:
Sure we can (I'll add it in the next commit)